### ISSUE TITLE
[TASK] Add Support for Governance of Protocol Substrate (Anchors, VAnchors, Signature Bridge)

### ISSUE THREAD

**Issue by shekohex** [webb-tools/relayer#187]
## Overview

Right now, the relayer has support for the following parts of Protocol Substrate:

- [x] Private transaction relaying (for Mixers, Anchors, and VAnchors)
- [x] Data querying (for zero-knowledge proof generation)
- [x] Event listening, proposing, and signature relaying.

that's the part we are currently missing, the Oracle Relayer over Protocol Substrate.

## Moving Forward and Tasks

We need to implement the following features, for `protocol-substrate`, preferably the same way we did for `protocol-solidity`.

1. A `{V}Anchor` Events Watcher that will create an `AnchorUpdateProposal` whenever their Merkle Tree gets updated (like in a deposit event).
2. A `SignatureBridge` Pallet events watcher that will receive the signed proposals to be executed whenever we see a proposal that targets this bridge.

### Data flow and implementation details

* **Example: Substrate-Substrate System**
	- User does a deposit on `{V}Anchor` `{ node: "foo", tree_id: 6 }` 
	- `{V}AnchorWatcher` see this event, and should try to update its own `linked_anchors`.
	- for every linked `{V}Anchor` it creates `AnchorUpdateProposal` (_the structure of this proposal depends on the target system_).
	- and for example, one of these linked `{V}Anchor`s is `{ node: "bar", tree_id: 8 }` so the relayer construct the following:
		```rust
		let target_chain_id = ctx.get_chain_id_of("bar");
		let target_chain = TypedChainId::Substrate(target_chain_id);
		let target_system = TargetSystem::new_tree_id(8);
        let resource_id = ResourceId::new(target_system, target_chain);
        let src_chain = TypedChainId::Substrate(my_chain_id);
		let target = [0u8; 32];
		let proposal = substrate::AnchorUpdateProposal::builder()
            .resource_id(resource_id)
            .src_chain(src_chain)
            .merkle_root(merkle_root)
            .latest_leaf_index(latest_leaf_index)
            .target(target)
            .build();
		```
	- Later, we send the `proposal` to be signed by one of the `ProposalSigningBackend`s
	- The Proposal Signing Backend signs the proposal (and _optionally_ sends it to the right `SignatureBridge` to execute it).
	- The `SignatureBridge` receives a `cmd` with `ExecuteProposalWithSignature` that contains the proposal `data` and the `signature`.
	- Lastly, the `SignatureBridge` verifies the signature, constructs the call/tx, and then sends it to the `TxQueue` to be executed.

* **Example: Any-Any System** (_future implementation_)
	- User does a deposit on `{V}Anchor` `{ node: "foo", tree_id: 6 }`, `{ chain: "polygon", address = "0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174" }`, _OR ANY OTHER SYSTEM_
	- `{V}AnchorWatcher` see this event, and should try to update its own `linked_anchors`.
	- for every linked `{V}Anchor` it creates `AnchorUpdateProposal` (_the structure of this proposal depends on the target system_).
	- and for example, these linked `{V}Anchor`s are `{ node: "webb", tree_id: 8 }` and `{ chain: "mainnet", address = "0x43bC9BFA99b4160d6aC972CE7249ee4ffBe3C13F" }` so the relayer construct the following:
		```rust
		// for webb substrate
		let target_chain_id = ctx.get_chain_id_of("webb");
		let target_chain = TypedChainId::Substrate(target_chain_id);
		let target_system = TargetSystem::new_tree_id(8);
        let resource_id = ResourceId::new(target_system, target_chain);
		// here the src is the foo node.
        let src_chain = TypedChainId::Substrate(my_chain_id);
		let target = [0u8; 32];
		let proposal = substrate::AnchorUpdateProposal::builder()
            .resource_id(resource_id)
            .src_chain(src_chain)
            .merkle_root(merkle_root)
            .latest_leaf_index(latest_leaf_index)
            .target(target)
            .build();
		```
		```rust
		// for mainnet evm chain
		let target_chain_id = 1;
		let target_chain = TypedChainId::Evm(target_chain_id);
		let target_system = TargetSystem::new_contract_address(linked_anchor.common.address);
        let resource_id = ResourceId::new(target_system, target_chain);
        let src_chain = TypedChainId::Substrate(my_chain_id);
		let target = [0u8; 32];
		let proposal = evm::AnchorUpdateProposal { resource_id, src_chain, merkle_root, leaf_index, target };
		```
	- Later, we send the `proposal` to be signed by one of the `ProposalSigningBackend`s
	- The Proposal Signing Backend signs the proposal (and _optionally_ sends it to the right `SignatureBridge` to execute it).
	- The `SignatureBridge` receives a `cmd` with `ExecuteProposalWithSignature` that contains the proposal `data` and the `signature`.
	- Lastly, the `SignatureBridge` verifies the signature, constructs the call/tx, and then sends it to the `TxQueue` to be executed. 

### Final Thoughts and Roadblocks

1. There is no notion of `SubstrateBridgeWatcher` in the relayer right now, it should be fairly easy, just a new trait that bounds to `SubstrateEventWatcher`, and this trait should be similar to `BridgeWatcher` one in the EVM.
2. The current Proposals Signing Backends are very opinionated around EVM based proposals, and to be honest ultimately the solution to this is the following:

```rust
// For MockedSigningBackend

#[async_trait::async_trait]
impl<P, C, S> super::ProposalSigningBackend<AnchorUpdateProposal>
    for MockedProposalSigningBackend<S, C>
where
	P: Into<Vec<u8>>, // anything we can convert to bytes.
	C: serde::Serialize, // any command we can send to the bridge.
    S: QueueStore<C, Key = SledQueueKey> + Send + Sync + 'static,
{
    async fn can_handle_proposal(
        &self,
        proposal: &P,
    ) -> anyhow::Result<bool> {
        // The Proposal Header is the first 40 bytes of each proposal.
        let proposal_header = header_from(proposal);
        let dest_chain_id = proposal_header.resource_id().typed_chain_id();
        let known_bridge = self.signature_bridges.contains_key(&dest_chain_id);
        Ok(known_bridge)
    }

    async fn handle_proposal(
        &self,
        proposal: &AnchorUpdateProposal,
    ) -> anyhow::Result<()> {
        // the way this one works is that we get the hash of the proposal bytes,
        // the we use the hash to be signed by the signer.
        // Read more here: https://bit.ly/3rqNYTU
		let proposal_header = header_from(proposal);
        let dest_chain_id = proposal_header.resource_id().typed_chain_id();
        let target_system = self.bridge_metadata(dest_chain_id)?;
        let signer = self.signer(dest_chain_id)?;
        let proposal_bytes = proposal.to_bytes();
        let hash = keccak256(&proposal_bytes);
        let signature = signer.sign_hash(H256::from(hash), false);
        let bridge_key = BridgeKey::new(target_system, dest_chain_id);
        tracing::debug!(
            %bridge_key,
            ?proposal,
            "Signaling Signature Bridge to execute proposal",
        );
        let signature_bytes = signature.to_vec();
        tracing::event!(
            target: crate::probe::TARGET,
            tracing::Level::DEBUG,
            kind = %crate::probe::Kind::SigningBackend,
            backend = "Mocked",
            signal_bridge = %bridge_key,
            data = ?hex::encode(&proposal_bytes),
            signature = ?hex::encode(&signature_bytes),
        );
        // now all we have to do is to send the data and the signature to the signature bridge.
        self.store.enqueue_item(
            SledQueueKey::from_bridge_key(bridge_key),
            BridgeCommand::ExecuteProposalWithSignature {
                data: proposal_bytes.to_vec(),
                signature: signature_bytes,
            },
        )?;
        Ok(())
    }
} 
```

3. There will be no need to introduce a `SubstrateBridgeCommand` as we can use the `BridgeCommand` and we actually could construct the call from the proposal data, we can detect the proposal type from the function signature. (cc @drewstone we will have to talk more about utilizing function signatures for substrate too.
4. The current implementation of `Substrate` Webb Proposals is _not_ quite correct, there is no separation between getting the underlying proposal data and constructing the call, the one we are missing is getting the proposal data.

Here is what we should update, for example, the `substrate::AnchorUpdateProposal` should look like the following:

```diff,rust
pub struct AnchorUpdateProposal {
    #[builder(default = 50)]
    pallet_index: u8,
    #[builder(default = 1)]
    call_index: u8,
-	resource_id: ResourceId,
+  header: ProposalHeader,
    src_chain: TypedChainId,
    merkle_root: [u8; 32],
    latest_leaf_index: u32,
    target: [u8; 32],
}
```

It looks the same as [`evm::AnchorUpdateProposal`](https://github.com/webb-tools/webb-rs/blob/37c86fa608809a27ed94562ac4d540e65f23b9d8/proposals/src/proposal/evm/anchor_update.rs#L20) and should be the same, and we should convert between both of them, except when we want to create the substrate call, so instead of this function here [`substrate::AnchorUpdateProposal::to_bytes`](https://github.com/webb-tools/webb-rs/blob/37c86fa608809a27ed94562ac4d540e65f23b9d8/proposals/src/proposal/substrate/anchor_update.rs#L61) returns the _call_ bytes, it should return the underlying proposal bytes, and we should have another method like `encoded_call_bytes` for example to return the call bytes.

5. There is no **"`linked_anchors`"** thing for substrate-based nodes configuration yet, as of now we watch the whole pallet events, so for the `AnchorBn254PalletConfig` we could add another configuration with `linked_anchors` similar to the one for evm-based ones defined [here](https://github.com/webb-tools/relayer/blob/6dc5b24353f9f029616c827234f29c146d5aa46d/src/config.rs#L314-L316). However, the `SubstrateLinkedAnchor`s one should be in this format `{ node = "node_name", tree_id = 10 }`.

---

> cc @dutterbutter could you also please compile this SPEC into smaller tasks and areas? since it touches quite a lot of things here in the relayer and also webb-rs.

**Comment by salman01zp** [comment]
@shekohex  Can you say more about generalizing `MockedProposalSigningBackend`? since it is also defined around only EVM.

```rust
/// A ProposalSigningBackend that uses the Governor's private key to sign proposals.
#[derive(TypedBuilder)]
pub struct MockedProposalSigningBackend<S>
where
    S: QueueStore<BridgeCommand, Key = SledQueueKey>,
{
    /// A map between chain id and its signature bridge system.
    #[builder(setter(into))]
    signature_bridges: HashMap<TypedChainId, TargetSystem>,
    /// Something that implements the QueueStore trait.
    store: Arc<S>,
    /// The private key of the governor.
    /// **NOTE**: This must be the same for all signature bridges.
    private_key: PrivateKey,
}
```

**Comment by shekohex** [comment]
> Also here. In protocol-substrate we have edgeList in pallet-linkable-tree, which will be empty initially. In order to add/update edge we have to create AnchorUpdateProposal

No need to read from the `edgeList`, we should use `linked-anchor`s in the configuration.

> Can you say more about generalizing MockedProposalSigningBackend? since it is also defined around only EVM.

Yes, that's the current implementation, the new modified one will take `Vec<u8>` (the full proposal bytes) instead of an actual proposal. that way it will work with all systems assuming each proposal will have the proposal header as the first 40 bytes.

**Comment by drewstone** [comment]
I think we should disambiguate if the `SignatureBridge` you keep referring to here is the smart contract or the event watcher. Maybe be more verbose and use `SignatureBridgeEventWatcher` so that all things remain especially clear to the reader.

**Comment by drewstone** [comment]
> Yes, that's the current implementation, the new modified one will take Vec<u8> (the full proposal bytes) instead of an actual proposal. that way it will work with all systems assuming each proposal will have the proposal header as the first 40 bytes.

This sounds like the best path forward too. The way _any_ signer should ultimately work is that it just acts as a signing entity or a delegator to a signing entity. It should not need to deal with actually formatting proposals, so that we may keep it as generalized as possible. The proposal creation should happen exclusively _before_ sending to the `SigningBackend` in my opinion.

**Comment by drewstone** [comment]
@salman01zp we will need to update the Substrate anchor proposal (and subsequently all other Substrate proposal types) to include the proposal header:
```
pub struct AnchorUpdateProposal {
    header: ProposalHeader,
    #[builder(default = 50)]
    pallet_index: u8,
    #[builder(default = 1)]
    call_index: u8,
    src_chain: TypedChainId,
    merkle_root: [u8; 32],
    latest_leaf_index: u32,
    target: [u8; 32],
}
```

**Comment by shekohex** [comment]
Closed in #188
